p2pserver/protocols/utils/parker.go

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package utils
+
+import "time"
+
+type Parker struct {
+	c chan struct{}
+}
+
+func NewParker() *Parker {
+	c := make(chan struct{}, 0)
+	return &Parker{c: c}
+}
+
+func (self *Parker) ParkTimeout(d time.Duration) {
+	select {
+	case <-self.c:
+	case <-time.After(d):
+	}
+}
+
+func (self *Parker) Unpark() {
```

Comment:
what happens if concurrent unpark?

Reply:
all operations are serialized by internal channel. If concurrent `unpark` happen before `park`, they all will be combined like one unpark, this behavior like [thread::park/unpark](https://doc.rust-lang.org/std/thread/fn.park.html) in rust.

Reply:
how are they combined?